src/main/java/com/ibm/cloud/sdk/core/util/CredentialUtils.java

```diff
@@ -108,84 +96,96 @@ public static boolean hasBadStartOrEndChar(String credentialValue) {
-          final JsonObject service = instance.getAsJsonObject();
-          final String instancePlan = service.get(PLAN).getAsString();
-          if ((plan == null) || plan.equalsIgnoreCase(instancePlan)) {
-            return instance.getAsJsonObject().getAsJsonObject(CREDENTIALS);
-          }
-        }
-      }
-    }
-    return null;
+  public static class VcapService {
+    public String name;
+    public String label;
+    public String plan;
+    public List<String> tags;
+    public VcapCredentials credentials;
   }
 
   /**
-   * Returns the property named 'key' associated with the specified service.
-   * @param serviceName the name of the service
-   * @param key the name of the property to retrieve
-   * @return the value of the specified property
+   * Retrieves the VCAP_SERVICES environment variable and unmarshals it.
+   * @return a map containing the unmarshalled VCAP_SERVICES value
    */
-  public static String getVcapValue(String serviceName, String key) {
-    return getVcapValue(serviceName, key, null);
+  public static Map<String, List<VcapService>> getVcapServicesObj() {
+    Map<String, List<VcapService>> result = null;
+
+    // Retrieve the environment variable's value.
+    String vcapValue = EnvironmentUtils.getenv(VCAP_SERVICES);
+    if (StringUtils.isNotEmpty(vcapValue)) {
+      Gson gson = GsonSingleton.getGson();
+      // Parse it into a map of VcapService lists keyed by service name.
+      Type typeToken = new TypeToken<Map<String, List<VcapService>>>() { }.getType();
+      try {
+        result = gson.fromJson(vcapValue, typeToken);
+      } catch (Throwable t) {
+        log.log(Level.WARNING, "Error parsing VCAP_SERVICES", t);
+      }
+    }
+    return result;
   }
 
-  /**
-   * Returns the value associated with the provided key from the VCAP_SERVICES, or null if it doesn't exist.
+   /**
+   * Returns an appropriate "service" list item for the specific 'serviceName' value.
    *
-   * @param serviceName the service name
-   * @param key the key whose value should be returned
-   * @param plan the plan name
-   * @return the value of the provided key
+   * @param serviceName the name used to locate the appropriate entry within the VCAP_SERVICES structure.
+   * This value will be used in two ways:
+   *
+   * First, a search will be done to find a service entry whose 'name' field matches the 'vcapName'
+   * parameter value.
+   * Second, if no entry with a matching 'name' field was found, then the first service entry within the
+   * services list whose key matches 'serviceName' will be returned.
+   *
+   *
+   * @return a VcapService instance if found, or null
    */
-  public static String getVcapValue(String serviceName, String key, String plan) {
-    if ((serviceName == null) || serviceName.isEmpty()) {
-      return null;
-    }
-
-    final JsonObject services = getVcapServices();
-    if (services == null) {
+  protected static VcapService getVcapServiceEntry(String serviceName) {
```

Comment:
@padamstx @mkistler - I've made this method protected vs public. Though I don't think users will actually be calling this function themselves. Please let me know otherwise